soccer/gameplay/situational_play_selection.py

```diff
@@ -0,0 +1,620 @@
+
+        if (self.currentPreempt):
+            self.currentPreempt = False
+        elif (self.situationChangeTime != None):
+            if (abs(time.time() - self.situationChangeTime) >
+                    self.playPreemptTime):
+                self.situationChangeTime = None
+                self.currentPreempt = True
+                self.LastSituation = self.currentSituaion
+
+    '''def addPreempt(play) possibly a function to add transition out of every state to the completed state with preemptPlay as the lambda
+        for g in states:
+            play.add_transition(g -> completed , preemptPlay)'''
+
+    #You will also need to make sure you delete all subbehaviors on enter_completed in the play
+
+    #A function to determine if the currently running play should be preempted 
+    def preemptPlay(self):
+        return self.currentPreempt
+
+    def ballToRobotDist(self, robot):
+        return math.sqrt((robot.pos.x - self.systemState.ball.pos.x)**2 + (
+            robot.pos.y - self.systemState.ball.pos.y)**2)
+
+    #Returns a tuple of the closest robot to the ball and the distance that robot is away from the ball
+
+    def closestRobot(self):
+        closestRobot = None
+        closestRobotDistance = 0.0
+        ballLocation = self.systemState.ball.pos
+        for g in self.activeRobots:
+            roboDist = self.ballToRobotDist(g)
+            if (closestRobot == None or roboDist < closestRobotDistance):
+                closestRobot = g
+                closestRobotDistance = roboDist
+
+        return (closestRobot, closestRobotDistance)
+
+    #Returns true if our robot is closest to the ball
+
+    def ourRobotClosest(self):
+        return self.closestRobot()[0].is_ours()
+
+    #Returns a tuple of the distance from the ball of our closest robot and our opponents closest robot
+    #(our distance, theirDistance)
+
+    def bothTeamsClosest(self):
+        ourClosest = None
+        ourClosestDist = 0.0
+        theirClosest = None
+        theirClosestDist = 0.0
+
+        ballLocation = self.systemState.ball.pos
+
+        for g in self.activeRobots:
+            roboDist = self.ballToRobotDist(g)
+            if (g.is_ours()):
+                if (ourClosest == None or roboDist < ourClosestDist):
+                    ourClosest = g
+                    ourClosestDist = roboDist
+            else:
+                if (theirClosest == None or roboDist < theirClosestDist):
+                    theirClosest = g
+                    theirClosestDist = roboDist
+
+        return (ourClosestDist, theirClosestDist)
+
+    #Returns the ratio of our closest distance to the ball and the opponents closest distance to the ball
+
+    def ballClosenessRatio(self):
+        distances = self.bothTeamsClosest()
+        try:
+            return distances[0] / distances[1]
+        except:
+            return 1.0
+
+    #Returns the robot that last had the ball and how long it was since they had the ball
+
```

Comment:
Why do you keep leaving these awkward lines between the comment and the function lmao